File: 4SumII/solution.java
class Solution {
  public int fourSumCount(int[] A, int[] B, int[] C, int[] D) {
    int count = 0;
    Map<Integer, Integer> sumsMap = this.calcSumsMap(A, B);
    for (int i : C) {
      for (int j : D) {
        int sum = -1 * (i + j);
        count += sumsMap.getOrDefault(sum, 0);
      }
    }
    return count;
  }

  private Map<Integer, Integer> calcSumsMap(int[] list1, int[] list2) {
    Map<Integer, Integer> sumsMap = new HashMap<>();
    for (int i: list1) {
      for (int j: list2) {
        int sum = i + j;
        sumsMap.put(sum, sumsMap.getOrDefault(sum, 0) + 1);
      }
    } 
   return sumsMap;
  }
}

File: 4SumII/solution.js
/**
 * @param {number[]} A
 * @param {number[]} B
 * @param {number[]} C
 * @param {number[]} D
 * @return {number}
 */
var fourSumCount = function(A, B, C, D) {
  const sumsMap = {};
  for (let i of A) {
    for (let j of B) {
      const sum = i + j;
      const count = sumsMap[sum] || 0;
      sumsMap[sum] = count + 1;
    }
  }
  let total = 0;
  for (let i of C) {
    for (let j of D) {
      const sum = - i - j;
      total += sumsMap[sum] || 0;
    }
  }
  return total;
};
